src/rpc/blockchain.cpp

```diff
@@ -1637,6 +1640,249 @@ static UniValue savemempool(const JSONRPCRequest& request)
+        std::vector<CPubKey> pubkeys;
+        UniValue pubkeys_uni = find_value(request.params[1], "pubkeys");
+        if (pubkeys_uni.isArray()) {
+            for (const UniValue& pubkey_uni : pubkeys_uni.get_array().getValues()) {
+                if (!pubkey_uni.isStr() || !IsHex(pubkey_uni.get_str())) {
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "public key must be hex encoded");
+                }
+                std::vector<unsigned char> data(ParseHex(pubkey_uni.get_str()));
+                CPubKey pub_key(data.begin(), data.end());
+                if (!pub_key.IsFullyValid()) {
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid public key");
+                }
+                pubkeys.push_back(pub_key);
+            }
+        }
+
+        // check for xpubs and derive a lookup window
+        UniValue xpubs_uni = find_value(request.params[1], "xpubs");
+        if (xpubs_uni.isArray()) {
+            for (const UniValue& xpub_uni : xpubs_uni.get_array().getValues()) {
+                if (!xpub_uni.isObject()) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid xpub object");
+                }
+
+                UniValue xpubkey_uni = find_value(xpub_uni, "xpub");
+                UniValue lookup_window_uni = find_value(xpub_uni, "lookupwindow");
+                //default lookup window
+                unsigned int window_start = 0;
+                unsigned int window_end = 1000;
+                if (!lookup_window_uni.isNull()) {
+                    if (lookup_window_uni.isArray() && lookup_window_uni.get_array().size() != 2) {
+                        throw JSONRPCError(RPC_INVALID_PARAMETER, "xpub lookupwindow must be an array with two elements (range)");
+                    }
+                    else {
+                        window_start = (unsigned int)lookup_window_uni.get_array().getValues()[0].get_int();
+                        window_end = (unsigned int)lookup_window_uni.get_array().getValues()[1].get_int();
+                    }
+                }
+                // Derive internal- and external-chain keys
+                CExtPubKey xpub = DecodeExtPubKey(xpubkey_uni.get_str());
+                if (!xpub.pubkey.IsValid()) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "invalid xpub");
+                }
+                CExtPubKey c0;
+                CExtPubKey c1;
+                xpub.Derive(c0, 0);
+                xpub.Derive(c1, 1);
+
+                for (unsigned int i = window_start; i <= window_end; i++) {
+                    // derive both (internal and external chain) child keys
+                    CExtPubKey k;
+                    c0.Derive(k, i);
+                    pubkeys.push_back(k.pubkey);
+                    c1.Derive(k, i);
+                    pubkeys.push_back(k.pubkey);
+                }
+            }
+        }
+
+        // add all common scripts for the given and derived pubkeys
```

Comment:
It's unfortunate that this introduces yet another "default set of addresses derived from a given key". Maintaining this will lead to incompatibilities between implementations as new types of scripts get added.

It would be better to have an explicit way to describe a set of scripts to watch for that could be reused in multiple places. I'm working towards that (see https://gist.github.com/sipa/125cfa1615946d0c3f3eec2ad7f250a2), but it may be a while before it's usable.

However, would you consider changing the RPC interface to be more easily adaptable to such a design? I think an interface which just takes a list of descriptions, each of which is a JSON object with either an address, script, or xpub (where the xpub one also takes a window size in a separate field). That means we can later add extra fields to describe what derivations to use, for example.